tar/tar.go

```diff
@@ -0,0 +1,178 @@
+	// Defaults to the current working directory.
+	rootDir string
+
+	// maxUntarSize represents the limit size for archives being decompressed by Untar.
+	// When max is equal or less than 0 disables size checks.
+	maxUntarSize int
+}
+
+// NewTar creates a new Tar.
+func NewTar(tarOpts ...TarOption) (*Tar, error) {
+	t := &Tar{}
+
+	if err := t.applyDefaults(); err != nil {
+		return nil, fmt.Errorf("cannot apply default values: %w", err)
+	}
+	if err := t.applyOpts(tarOpts...); err != nil {
+		return nil, fmt.Errorf("cannot apply tarOpts: %w", err)
+	}
+
+	return t, nil
+}
+
+// Untar reads the gzip-compressed tar file from r and writes it into dir.
+// If dir is a relative path, it will always descend from rootDir, which
+// defaults to the working dir.
+func (t *Tar) Untar(r io.Reader, dir string) (summary string, err error) {
+	dir = filepath.Clean(dir)
+	if !filepath.IsAbs(dir) {
+		dir, err = securejoin.SecureJoin(t.rootDir, dir)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	fi, err := os.Lstat(dir)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return "", fmt.Errorf("cannot lstat '%s': %w", dir, err)
+	}
+
+	if err == nil && !fi.IsDir() {
+		return "", fmt.Errorf("dir must be a directory: %w", err)
+	}
+
+	t0 := time.Now()
+	nFiles := 0
+	madeDir := map[string]bool{}
+	defer func() {
+		td := time.Since(t0)
+		if err == nil {
+			summary = fmt.Sprintf("Extracted tarball into %s: %d files, %d dirs (%v)", dir, nFiles, len(madeDir), td)
+		} else {
+			summary = fmt.Sprintf("Error extracting tarball into %s after %d files, %d dirs, %v: %v", dir, nFiles, len(madeDir), td, err)
+		}
+	}()
+	zr, err := gzip.NewReader(r)
+	if err != nil {
+		return summary, fmt.Errorf("requires gzip-compressed body: %w", err)
+	}
+	tr := tar.NewReader(zr)
+	loggedChtimesError := false
+	processedBytes := 0
+	for {
+		f, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			log.Printf("tar reading error: %v", err)
+			return summary, fmt.Errorf("tar error: %w", err)
+		}
+		processedBytes += int(f.Size)
+		if t.maxUntarSize > 0 && processedBytes > t.maxUntarSize {
+			return summary, fmt.Errorf("tar %q is bigger than max archive size of %d bytes", f.Name, t.maxUntarSize)
+		}
+		if !validRelPath(f.Name) {
+			return summary, fmt.Errorf("tar contained invalid name error %q", f.Name)
+		}
+		rel := filepath.FromSlash(f.Name)
+		abs := filepath.Join(dir, rel)
+
+		fi := f.FileInfo()
+		mode := fi.Mode()
+		switch {
+		case mode.IsRegular():
+			// Make the directory. This is redundant because it should
+			// already be made by a directory entry in the tar
+			// beforehand. Thus, don't check for errors; the next
+			// write will fail with the same error.
+			dir := filepath.Dir(abs)
+			if !madeDir[dir] {
+				if err := os.MkdirAll(filepath.Dir(abs), 0o755); err != nil {
+					return summary, err
+				}
+				madeDir[dir] = true
+			}
+			wf, err := os.OpenFile(abs, os.O_RDWR|os.O_CREATE|os.O_TRUNC, mode.Perm())
```

Comment:
Looks like upstream has some updated around this code https://github.com/golang/build/blame/master/internal/untar/untar.go#L82-L93 . Since this is used in flux CLI which can be used on macOS, do we need the updates here too?